### ISSUE TITLE
Scrobbling error on latest commit

### ISSUE THREAD

**Issue by stebe** [ampache/ampache#965]
I just pulled commit 1446bd22c6432cfae04c71af83d3d60fa0be999a and I'm getting this error when scrobbling to libre.fm. 

 Registering stream stats for {Toshihiko Sahashi - TEMA III}... 
2015-09-22 17:45:38 [admin] (user.class.php) -> Updating stats for {song/309950} {Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/45.0.2454.93 Safari/537.36}... 
2015-09-22 17:45:38 [ampache] (PHP) -> [Runtime Error] fopen(): SSL operation failed with code 1. OpenSSL Error messages: 
2015-09-22 17:45:38 [ampache] (PHP) -> error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed in file /usr/share/ampache/stebe-git/ampache/lib/class/scrobbler.class.php(97) 
2015-09-22 17:45:38 [ampache] (PHP) -> [Runtime Error] fopen(): Failed to enable crypto in file /usr/share/ampache/stebe-git/ampache/lib/class/scrobbler.class.php(97) 
2015-09-22 17:45:38 [ampache] (PHP) -> [Runtime Error] fopen(https://libre.fm/2.0/): failed to open stream: operation failed in file /usr/share/ampache/stebe-git/ampache/lib/class/scrobbler.class.php(97) 
2015-09-22 17:45:38 [admin] (Libre.FM) -> Error Submit Failed: Did not receive a valid response

**Comment by stebe** [comment]
From my debug screen.

ibrefm_user	steblublu
librefm_md5_pass	removed
librefm_port	80
librefm_host	turtle.libre.fm
librefm_url	/submissions/1.2/
librefm_challenge	removed

**Comment by nioc** [comment]
Hi, may be it is my fault :disappointed: 

Do you have updated the module from plugin screen ?
Can you give the result of the following SQL request :
`SELECT * FROM ampache.preference where name like '%fm%' order by id desc`

You should have 4 lines :
`# id, name, value, description, level, type, catagory
'166', 'librefm_grant_link', '', 'Libre.FM Grant URL', '25', 'string', 'plugins'
'165', 'lastfm_grant_link', '', 'Last.FM Grant URL', '25', 'string', 'plugins'
'164', 'librefm_challenge', '', 'Libre.FM Submit Challenge', '25', 'string', 'internal'
'158', 'lastfm_challenge', '', 'Last.FM Submit Challenge', '25', 'string', 'internal'`

**Comment by stebe** [comment]
Here it is...

MariaDB [ampache-doped]> SELECT * FROM ampache.preference where name like '%fm%' order by id desc;
 +----+------------------+-------+--------------------------+-------+--------+----------+
 | id | name             | value | description              | level | type   | catagory |
+----+------------------+-------+--------------------------+-------+--------+----------+
| 78 | lastfm_challenge |       | Last.FM Submit Challenge |    25 | string | internal |
| 77 | lastfm_url       |       | Last.FM Submit URL       |    25 | string | internal |
| 76 | lastfm_host      |       | Last.FM Submit Host      |    25 | string | internal |
| 75 | lastfm_port      |       | Last.FM Submit Port      |    25 | string | internal |
| 74 | lastfm_pass      |       | Last.FM Password         |    25 | string | plugins  |
| 73 | lastfm_user      |       | Last.FM Username         |    25 | string | plugins  |
+----+------------------+-------+--------------------------+-------+--------+----------+
6 rows in set (0.01 sec)

**Comment by nioc** [comment]
Ok, as I said, following lines should be deleted :
`librefm_user`
`librefm_md5_pass`
`librefm_url`
`librefm_host`
`librefm_port`

See the [Librefm.plugin.php] (https://github.com/ampache/ampache/blob/b336d1ab7e0c2e7bfa8ce8ccc2659252b578a2fa/modules/plugins/Librefm/Librefm.plugin.php) at line 87 : 
````
    public function upgrade()
    {
        $from_version = Plugin::get_plugin_version($this->name);
        if ($from_version < 2) {
            Preference::rename('librefm_pass', 'librefm_md5_pass');
        }
        if ($from_version < 3) {
            Preference::delete('librefm_md5_pass');
            Preference::delete('librefm_user');
            Preference::delete('librefm_url');
            Preference::delete('librefm_host');
            Preference::delete('librefm_port');
            Preference::insert('librefm_grant_link','Libre.FM Grant URL','','25','string','plugins');
        }
        return true;
    } // upgrade
````

Can you go on this url : /admin/modules.php?action=show_plugins
You should see :
`Libre.FM	Records your played songs to your Libre.FM Account	000003	000002	Update`
Click on 'Update'.

Then go on this : /preferences.php?tab=plugins
Click on the button at the right of "Libre.FM Grant URL" for authentication in libre.fm website and it should works... otherwise tell me, and I'll have a look on it.

**Comment by stebe** [comment]
I did do the update:
Libre.FM	Records your played songs to your Libre.FM Account	000003	000003	Deactivate

I go to the preferences:
Libre.FM Grant URL	Click for grant Ampache to Librefm

I'm sent to the access page:
Hello steblublu
Unknown client with
API key: 
Callback URL: https://ca.animeshare.com/ampache/preferences.php?tab=plugins&action=grant&plugin=Librefm
wants your permission to talk with this service.

Submit


I click on submit and I get this error when I'm back to ampache:

Error

Your Account has not been updated : Librefm 
 Continue

**Comment by nioc** [comment]
This is strange. I need to sleep (1 am for me), I'll have a look tomorrow.
Just in case, check with a firebug tool, if you get a token in callback url.

**Comment by stebe** [comment]
Looks like the token is being returned in firebug, I'm attaching the ampache debug log. I'm thinking it's an issue with SSL on with my version of php?  PHP 5.6.12-1



2015-09-22 19:54:10 [admin] (session) -> qsrjdbpgmml4rmjimv4rmae7m3 has been extended to Wed, 23 Sep 2015 19:54:10 -0400 extension length 86400 
2015-09-22 19:54:10 [ampache] (PHP) -> [Runtime Error] fopen(): SSL operation failed with code 1. OpenSSL Error messages: 
2015-09-22 19:54:10 [ampache] (PHP) -> error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed in file /usr/share/ampache/stebe-git/ampache/lib/class/scrobbler.class.php(97) 
2015-09-22 19:54:10 [ampache] (PHP) -> [Runtime Error] fopen(): Failed to enable crypto in file /usr/share/ampache/stebe-git/ampache/lib/class/scrobbler.class.php(97) 
2015-09-22 19:54:10 [ampache] (PHP) -> [Runtime Error] fopen(https://libre.fm/2.0/?method=auth.getSession&amp;api_key=&amp;token=b22de978fb8ac8dba2cc0f37d1772385&amp;api_sig=9ff270e67f3ccb91622a8e96cf2de98e): failed to open stream: operation failed in file /usr/share/ampache/stebe-git/ampache/lib/class/scrobbler.class.php(97) 
2015-09-22 19:54:10 [admin] (Libre.FM) -> getSession Failed: Did not receive a valid response 
2015-09-22 19:54:10 [admin] (session) -> Session created: 70b96a0b275cc574772e8e34dfe1ebbb 
2015-09-22 19:54:11 [admin] (Core) -> Registered post form confirmation with SID e43e1198d559f6e6e4943e750ce34f15 and expiration 1443052451 (86400 seconds from now) 
2015-09-22 19:54:11 [admin] (session) -> qsrjdbpgmml4rmjimv4rmae7m3 has been extended to Wed, 23 Sep 2015 19:54:11 -0400 extension length 86400 
2015-09-22 19:54:12 [admin] (session) -> qsrjdbpgmml4rmjimv4rmae7m3 has been extended to Wed, 23 Sep 2015 19:54:12 -0400 extension length 86400 
2015-09-22 19:54:12 [admin] (ajax.server.php) -> Called for page: {stats}

**Comment by Afterster** [comment]
For me it looks like you have openssl support but the certificate used by LibreFM cannot be verified successfully. It could be because the root certificate of LibreFM certificate is not approved on your server.

**Comment by stebe** [comment]
I'm using the default certificates installed by Debian and I have no problem with ssl connections to libre.fm with either firefox or chrome.

var_dump(openssl_get_cert_locations());

array (size=8)
  'default_cert_file' => string '/usr/lib/ssl/cert.pem' (length=21)
  'default_cert_file_env' => string 'SSL_CERT_FILE' (length=13)
  'default_cert_dir' => string '/usr/lib/ssl/certs' (length=18)
  'default_cert_dir_env' => string 'SSL_CERT_DIR' (length=12)
  'default_private_dir' => string '/usr/lib/ssl/private' (length=20)
  'default_default_cert_area' => string '/usr/lib/ssl' (length=12)
  'ini_cafile' => string '' (length=0)
  'ini_capath' => string '' (length=0)

**Comment by nioc** [comment]
It seems to be an OpenSSL bug with PHP 5.6 indeed.
You can edit the [Librefm.plugin.php] (https://github.com/ampache/ampache/blob/b336d1ab7e0c2e7bfa8ce8ccc2659252b578a2fa/modules/plugins/Librefm/Librefm.plugin.php) at line 39 to replace https by http : 
`private $scheme     = 'https';`
to this :
`private $scheme     = 'http';`

**Comment by stebe** [comment]
I just tried the change but the connection is still attempting an SSL connection. I think libre.fm forces a redirect to a secure connection. I tried with IE, Firefox and Chrome with all add-ons disabled and when I enter http://libre.fm the connection reloads as https. The problem is not with Ampache but with php and/or openssl.

**Comment by nioc** [comment]
Ok. Perhaps using curl may resolve your issue (see http://stackoverflow.com/a/14078318)?

**Comment by stebe** [comment]
Could the ssl configuration of the server affect this?

<IfModule mod_ssl.c>

	# Pseudo Random Number Generator (PRNG):
	# Configure one or more sources to seed the PRNG of the SSL library.
	# The seed data should be of good random quality.
	# WARNING! On some platforms /dev/random blocks if not enough entropy
	# is available. This means you then cannot use the /dev/random device
	# because it would lead to very long connection times (as long as
	# it requires to make more entropy available). But usually those
	# platforms additionally provide a /dev/urandom device which doesn't
	# block. So, if available, use this one instead. Read the mod_ssl User
	# Manual for more details.
	#
	SSLRandomSeed startup builtin
	SSLRandomSeed startup file:/dev/urandom 512
	SSLRandomSeed connect builtin
	SSLRandomSeed connect file:/dev/urandom 512

	##
	##  SSL Global Context
	##
	##  All SSL configuration in this context applies both to
	##  the main server and all SSL-enabled virtual hosts.
	##

	#
	#   Some MIME-types for downloading Certificates and CRLs
	#
	AddType application/x-x509-ca-cert .crt
	AddType application/x-pkcs7-crl	.crl

	#   Pass Phrase Dialog:
	#   Configure the pass phrase gathering process.
	#   The filtering dialog program (`builtin' is a internal
	#   terminal dialog) has to provide the pass phrase on stdout.
	SSLPassPhraseDialog  builtin

	#   Inter-Process Session Cache:
	#   Configure the SSL Session Cache: First the mechanism 
	#   to use and second the expiring timeout (in seconds).
	#   (The mechanism dbm has known memory leaks and should not be used).
	#SSLSessionCache		 dbm:${APACHE_RUN_DIR}/ssl_scache
	SSLSessionCache		shmcb:${APACHE_RUN_DIR}/ssl_scache(512000)
	SSLSessionCacheTimeout  300

	#   Semaphore:
	#   Configure the path to the mutual exclusion semaphore the
	#   SSL engine uses internally for inter-process synchronization. 
	#   (Disabled by default, the global Mutex directive consolidates by default
	#   this)
	#Mutex file:${APACHE_LOCK_DIR}/ssl_mutex ssl-cache


	#   SSL Cipher Suite:
	#   List the ciphers that the client is permitted to negotiate. See the
	#   ciphers(1) man page from the openssl package for list of all available
	#   options.
	#   Enable only secure ciphers:
	SSLCipherSuite         ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA
	
	# SSL server cipher order preference:
	# Use server priorities for cipher algorithm choice.
	# Clients may prefer lower grade encryption.  You should enable this
	# option if you want to enforce stronger encryption, and can afford
	# the CPU cost, and did not override SSLCipherSuite in a way that puts
	# insecure ciphers first.
	# Default: Off
	SSLHonorCipherOrder on

#	SSLOpenSSLConfCmd DHParameters "/etc/ssl/private/dhparams.pem"

	#   The protocols to enable.
	#   Available values: all, SSLv3, TLSv1, TLSv1.1, TLSv1.2
	#   SSL v2  is no longer supported
	SSLProtocol all -SSLv2 -SSLv3

	#   Allow insecure renegotiation with clients which do not yet support the
	#   secure renegotiation protocol. Default: Off
	#SSLInsecureRenegotiation on

	#   Whether to forbid non-SNI clients to access name based virtual hosts.
	#   Default: Off
	#SSLStrictSNIVHostCheck On

	SSLCompression Off

</IfModule>

**Comment by stebe** [comment]
After some playing around and manually trying to run curl I was getting this error:
curl  https://libre.fm
curl: (77) Problem with the SSL CA cert (path? access rights?)

Found that the the /etc/ssl/certs folder was not exec so the certificate authorities could not be read. This was even after I purged and re-installed the ca-certificates package.
drwxr--r-- 3 root root      24576 Sep 23 08:16 certs
changed the permissions and I can now get the token and scrobbles are being sent.
drwxr-xr-x 3 root root      24576 Sep 23 08:16 certs

Sorry for wasting your time @nioc. I don't know how the permissions of that folder got changed.
